### ISSUE TITLE
Consider using temporary files on create, for tighter zstd control.

### ISSUE THREAD

**Issue by dholth** [conda/conda-package-handling#171]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Consider using temporary (possibly zstd -3 compressed) `info-` and `pkg-` tarballs when creating `.conda`.

According to the `zstd` documentation, zstandard is more efficient when you can tell the compressor the total size ahead of time. Maybe it will avoid allocating extra memory, to further improve on #167 ? We would write the temporary file, and then compress it into the ZIP-format `.conda` archive giving the total size to the compressor. We would need to do some experiments to verify that this is an improvement.

Unlike unpacking, the extra time spent creating these temporary files would usually be dwarfed by the time spent running the compressor at a high level.

**Comment by dholth** [comment]
One question is how bad is it if the size hint is wrong. And does the size hint really help?

I'm not sure whether or not we want to run tarfile.py twice over the same files, it would be worth measuring.

On my laptop, even zstd -1 compression is slower than the disk. But it is really nice to use less disk space, and not everyone will have a fast disk.

**Comment by dholth** [comment]
This suite https://github.com/conda/conda-benchmarks/blob/main/benchmarks/peakmem_tests.py is an easy way to check memory, and could produce graphs like this if we set it up. https://dholth.github.io/conda-benchmarks/#conda_install.TimeInstall.time_explicit_install?conda-package-handling=1.9.0&conda-package-handling=2.0.0a2&p-latency=0.0&p-threads=1

**Comment by dholth** [comment]
I wrote a test at https://gist.github.com/dholth/0a8b26ddd361ae9a2440f19ceceaaa2e

Compression takes about the same amount of time when the size is known. The compressor produces the same number of bytes, plus a 3-byte header, for this 10MB file.

If you give the compressor the wrong number of bytes, it just fails.

If the size was not known on compression, you have to use the streaming decompression API instead of the one-shot `.decompress(bytes)` API.

According to the scalene profiler, at level 22,

```
peak memory,  function
267MB oneshot
777MB onestream
266MB rightsize
774MB multistream

decompression peak memory
  9.9MB one-shot decompression
128.5MB streaming decompression, size unknown
 19.3MB streaming decompression, size known
(fails) one-shot decompression, size unknown
```

**Comment by dholth** [comment]
Fixed in #190 using the throwaway-TarFile strategy.

**Comment by dholth** [comment]
See also https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile
